File: 1/AI1110/Assignment-2/codes/icse12-2019-1-2.c
/*
ICSE 2019 Grade 12
Question 1(ii)

Name: Ankit Saha
Roll number: AI21BTECH11004

Problem Statement
Solve: sin(2arctan(x)) = 1
*/

#include <stdio.h>
#include <math.h>

// Function prototypes
float f(float x);
void writeCoords();

int main() {
    float x = 1.0;
    if (f(x) == 1.0) {
        printf("x = %g is a solution\n", x);
    } else {
        printf("x = %g is not a solution\n", x);
    }
    writeCoords();
}

// Returns the value of the given function at x
float f(float x) {
    return sin(2 * atan(x));
}

// Writes the x and y coordinates into a .dat file
void writeCoords() {
    FILE *fp = fopen("points.dat", "w");
    if (!fp) {
        printf("Couldn't open file\n");
        return;
    }
    for (float i = -6; i <= 6; i += 0.01) {
        fprintf(fp, "%f %f\n", i, f(i));
    }
    fclose(fp);
    return;
}

File: 1/AI1110/Assignment-2/codes/icse12-2019-1-2.py
# ICSE 2019 Grade 12
# Question 1(ii)

# Name: Ankit Saha
# Roll number: AI21BTECH11004

""" Problem Statement
Solve: sin(2arctan(x)) = 1
"""

import numpy as np
import matplotlib.pyplot as plt

# Preparing to plot y = sin(2arctan(x))
coords = np.loadtxt('points.dat', dtype='float')
X = []
Y = []
for coord in coords:
    X.append(coord[0])
    Y.append(coord[1])

# Preparing to plot y = 1
X2 = np.linspace(-6, 6, 1200)
Y2 = np.linspace(1, 1, 1200)

# Plotting
plt.plot(X, Y, label='$y = \sin(2\\tan^{-1}x)$')
plt.plot(X2, Y2, label='$y = 1$')

# Displaying the solution
sol = np.zeros((2,1))
sol[0][0] = 1
sol[1][0] = 1
A = sol[0]
B = sol[1]
plt.plot(A, B, 'o')
for xy in zip(A, B):
    plt.annotate('(%s, %s)' %xy, xy=xy, xytext=(30, 0), textcoords='offset points')

plt.grid()
plt.legend(loc='best', prop={'size':11})
plt.xlabel('$x$')
plt.ylabel('$y$')
plt.show()